sys/linux/ksmbd.txt

```diff
@@ -0,0 +1,168 @@
+include <linux/ip.h>
+include <linux/ipv6.h>
+include <linux/route.h>
+include <uapi/linux/if_arp.h>
+include <uapi/linux/netfilter_ipv6/ip6_tables.h>
+include <uapi/linux/wireless.h>
+include <uapi/linux/in.h>
+
+sync_id {
+	ProcessId	int32
+	TreeId		int32
+} [packed]
+
+_union_smb_req_id [
+	sync_id		sync_id
+	async_id	int64
+]
+
+negotiate_message {
+	Signature			array[int8, 8]
+#  NtLmNegotiate = 1
+	MessageType			int32
+	NegotiateFlags			int32
+# MBZ unless SMB3.02 or later
+	DomainName_Length		len[DomainString, int16]
+	DomainName_MaximumLength	len[DomainString, int16]
+	DomainName_Offset		offsetof[DomainString, int32]
+# RFC 1001 and ASCII
+	WorkstationName_Length		len[Workstationname_Buffer, int16]
+	WorkstationName_MaximumLength	len[Workstationname_Buffer, int16]
+	WorkstationName_Offset		offsetof[Workstationname_Buffer, int32]
+
+# struct security_buffer for version info not present since we
+# do not set the version is present flag
+	DomainString			string
+
+# followed by WorkstationString
+	Workstationname_Buffer		string
+} [packed]
+
+_smb2_hdr_pre {
+	ProtocolId	const[0xfe534d42, int32be]
+	StructureSize	const[0x40, int16]
+	CreditCharge	int16
+	Status		int32
+} [packed]
+
+_smb2_hdr_post {
+	CreditRequest	int16
+	Flags		int32
+	NextCommand	int32
+	MessageId	int64
+	Id		_union_smb_req_id
+	SessionId	int64
+	Signature	array[int8, 16]
+} [packed]
+
+#_smb3_hdr_pre {
+#	ProtocolId	const[0xfe534d42, int32be]
+#	StructureSize	const[0x40, int16]
+#	CreditCharge	int16
+#	ChannelSequence	int16
+#	Reserved	int16
+#} [packed]
+
+#_smb3_hdr_post {
+#	CreditRequest	int16
+#	Flags		int32
+#	NextCommand	int32
+#	MessageId	int64
+#	Id		_union_smb_req_id
+#	SessionId	int64
+#	Signature	array[int8, 16]
+#} [packed]
```

Comment:
If these structs are not necessary, let's just drop them from the final descriptions.